pyuvdata/utils.py

```diff
@@ -5706,88 +5742,112 @@ def _get_dset_shape(dset, indices):
+        eval_ind = indices
     # assert indices is longer than 2, or return trivial solutions
-    if len(indices) == 0:
+    if len(eval_ind) == 0:
         return [slice(0, 0, 0)], False
-    elif len(indices) == 1:
-        return [slice(indices[0], indices[0] + 1, 1)], True
-    elif len(indices) == 2:
-        return [slice(indices[0], indices[1] + 1, indices[1] - indices[0])], True
+    if len(eval_ind) <= 2:
+        return [
+            slice(eval_ind[0], eval_ind[-1] + 1, max(eval_ind[-1] - eval_ind[0], 1))
+        ], True
+
+    # Catch the simplest case of "give me a single slice or exit"
+    if (max_nslice == 1) and return_index_on_fail:
+        step = eval_ind[1] - eval_ind[0]
+        if all(np.diff(eval_ind) == step):
+            return [slice(eval_ind[0], eval_ind[-1] + 1, step)], True
+        return [indices], False
 
     # setup empty slices list
-    Ninds = len(indices)
+    Ninds = len(eval_ind)
     slices = []
 
     # iterate over indices
-    for i, ind in enumerate(indices):
-        if i == 0:
-            # start the first slice object
-            start = ind
-            last_step = indices[i + 1] - ind
+    start = last_step = None
+    for ind in eval_ind:
+        if last_step is None:
+            # Check if this is the first slice, in which case start is None
+            if start is None:
+                start = ind
+                continue
+            last_step = ind - start
+            last_ind = ind
             continue
 
         # calculate step from previous index
-        step = ind - indices[i - 1]
+        step = ind - last_ind
 
         # if step != last_step, this ends the slice
         if step != last_step:
             # append to list
-            slices.append(slice(start, indices[i - 1] + 1, last_step))
-
-            # check if this is the last element
-            if i == Ninds - 1:
-                # append last element
-                slices.append(slice(ind, ind + 1, 1))
-                continue
+            slices.append(slice(start, last_ind + 1, last_step))
 
             # setup next step
             start = ind
-            last_step = indices[i + 1] - ind
+            last_step = None
+
+        last_ind = ind
 
-        # check if this is the last element
-        elif i == Ninds - 1:
-            # end slice and append
-            slices.append(slice(start, ind + 1, step))
+    # Append the last slice
+    slices.append(slice(start, ind + 1, last_step))
 
-    # determine whether slices are a reasonable representation
-    Nslices = len(slices)
-    passed = (float(Nslices) / len(indices)) < max_nslice_frac
+    # determine whether slices are a reasonable representation, and determine max_nslice
+    # if only max_nslice_frac was supplied.
+    if max_nslice is None:
+        max_nslice = max_nslice_frac * Ninds
+    check = len(slices) <= max_nslice
 
-    return slices, passed
+    return [indices] if (not check and return_index_on_fail) else slices, check
```

Comment:
I find this (and generally these kinds of in-line if statements) a little hard to read. I'd slightly prefer the longer multi-line if statement. Open to other votes however.

Reply:
Changed in 9c862da2